pallets/data-preservers/src/lib.rs

```diff
@@ -130,6 +172,84 @@ pub mod pallet {
         ValueQuery,
     >;
 
+    #[pallet::storage]
+    pub type Profiles<T: Config> =
+        StorageMap<_, Blake2_128Concat, ProfileId, RegisteredProfile<T>, OptionQuery>;
+
+    #[pallet::storage]
+    pub type NextProfileId<T: Config> = StorageValue<_, ProfileId, ValueQuery>;
+
+    /// Balance used by this pallet
+    pub type BalanceOf<T> =
+        <<T as Config>::Currency as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+    pub type ProfileId = u64;
+
+    /// Data preserver profile.
+    #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+    #[derive(
+        RuntimeDebugNoBound, PartialEqNoBound, EqNoBound, Encode, Decode, CloneNoBound, TypeInfo,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub struct Profile<T: Config> {
+        pub url: BoundedVec<u8, T::MaxBootNodeUrlLen>,
+        pub limited_to_para_ids: Option<BoundedVec<ParaId, T::MaxParaIdsVecLen>>,
+        pub mode: ProfileMode,
+    }
+
+    #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+    #[derive(RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo)]
+    pub enum ProfileMode {
+        Bootnode,
+        Rpc { supports_ethereum_rpcs: bool },
+    }
+
+    /// Profile with additional data:
+    /// - the account id which created (and manage) the profile
+    /// - the amount deposited to register the profile
+    #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+    #[derive(
+        RuntimeDebugNoBound, PartialEqNoBound, EqNoBound, Encode, Decode, CloneNoBound, TypeInfo,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub struct RegisteredProfile<T: Config> {
+        pub account: T::AccountId,
+        pub deposit: BalanceOf<T>,
+        pub profile: Profile<T>,
+    }
+
+    pub trait ProfileDeposit<Profile, Balance> {
+        fn profile_deposit(profile: &Profile) -> Result<Balance, DispatchErrorWithPostInfo>;
+    }
+
+    pub struct BytesProfileDeposit<BaseCost, ByteCost>(PhantomData<(BaseCost, ByteCost)>);
```

Comment:
I think this is not a bad idea to be honest, but it's the first time that I see it in the pallet haha. I like it but since it's a configuration option maybe we can move it to a different file?

Reply:
Also let's add some comments to it